content/helpers.go

```diff
@@ -230,8 +230,40 @@ func seekReader(r io.Reader, offset, size int64) (io.Reader, error) {
 }
 
 func copyWithBuffer(dst io.Writer, src io.Reader) (written int64, err error) {
-	buf := bufPool.Get().(*[]byte)
-	written, err = io.CopyBuffer(dst, src, *buf)
-	bufPool.Put(buf)
+	// If the reader has a WriteTo method, use it to do the copy.
+	// Avoids an allocation and a copy.
+	if wt, ok := src.(io.WriterTo); ok {
+		return wt.WriteTo(dst)
+	}
+	// Similarly, if the writer has a ReadFrom method, use it to do the copy.
+	if rt, ok := dst.(io.ReaderFrom); ok {
+		return rt.ReadFrom(src)
+	}
+	bufRef := bufPool.Get().(*[]byte)
+	defer bufPool.Put(bufRef)
+	buf := *bufRef
+	for {
+		nr, er := io.ReadAtLeast(src, buf, len(buf))
+		if nr > 0 {
+			nw, ew := dst.Write(buf[0:nr])
+			if nw > 0 {
+				written += int64(nw)
+			}
+			if ew != nil {
+				err = ew
+				break
+			}
+			if nr != nw {
+				err = io.ErrShortWrite
+				break
+			}
+		}
+		if er != nil {
```

Comment:
if you move this up before the write.. you can probably just use err instead of err, er, ew
and specify nw,nr as int vars above the while not done for loop

Reply:
FWIW this is the exact implementation of the io.CopyBuffer , with just the `Read` substituted with `ReadAtLeast` with the additional handling of `io.ErrUnexpectedEOF` I wanted to keep the implantation remarkably similar to the `io` implementation

Reply:
I see.. pls add a comment to the block of code stating what you just said :-)

Reply:
See https://golang.org/src/io/io.go?s=13435:13513#L387

Reply:
Sure, sorry missed your comment